
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Session 2: Track Alignment and Data Loading &#8212; Gos Tutorial</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Session 3: Semantic Zooming" href="03-Semantic-Zooming.html" />
    <link rel="prev" title="Session 1: Single Track" href="01-Single-Track.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/gosling-logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Gos Tutorial</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../README.html">
                    Genomics Data Visualization with Gos
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Index.html">
   Outline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="01-Single-Track.html">
   Session 1: Single Track
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Session 2: Track Alignment and Data Loading
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="03-Semantic-Zooming.html">
   Session 3: Semantic Zooming
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="04-Coordinated-Multiple-Views.html">
   Session 4: Coordinated Multiple Views
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/gosling-lang/gos-tutorial/master?urlpath=tree/notebooks/02-Track-Alignment-and-Data-Loading.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/notebooks/02-Track-Alignment-and-Data-Loading.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Session 2: Track Alignment and Data Loading
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#track-composition">
   Track Composition
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#layout-linear-and-circular">
     <code class="docutils literal notranslate">
      <span class="pre">
       layout
      </span>
     </code>
     - Linear and Circular
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#linear-view">
       Linear view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#circular-view">
       Circular view
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#alignment-multiple-tracks-in-one-view">
     <code class="docutils literal notranslate">
      <span class="pre">
       alignment
      </span>
     </code>
     -  Multiple Tracks in One View​
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#stacked-tracks-in-single-view">
       Stacked tracks in single view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#overlay-tracks-in-single-view">
       Overlay tracks in single view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#example-gene-annotation">
       Example: Gene Annotation
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#arrangement-arrange-multiple-views">
     <code class="docutils literal notranslate">
      <span class="pre">
       arrangement
      </span>
     </code>
     - Arrange Multiple Views​
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gos-horizontal-gos-vertical">
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.horizontal
        </span>
       </code>
       &amp;
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.vertical
        </span>
       </code>
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gos-serial-gos-parallel">
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.serial
        </span>
       </code>
       &amp;
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.parallel
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-data">
   Loading data
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-visualization">
     The visualization
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-dataset">
     The dataset
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remote-dataset-via-url">
     Remote dataset (via URL)
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#local-dataset-via-local-filepath">
     Local Dataset (via local filepath)
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#in-memory-via-pd-dataframe">
     In memory (via
     <code class="docutils literal notranslate">
      <span class="pre">
       pd.DataFrame
      </span>
     </code>
     )
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bonus-exercise-local-bigwig-track">
   (BONUS) Exercise: local bigwig track
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Session 2: Track Alignment and Data Loading</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Session 2: Track Alignment and Data Loading
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#track-composition">
   Track Composition
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#layout-linear-and-circular">
     <code class="docutils literal notranslate">
      <span class="pre">
       layout
      </span>
     </code>
     - Linear and Circular
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#linear-view">
       Linear view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#circular-view">
       Circular view
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#alignment-multiple-tracks-in-one-view">
     <code class="docutils literal notranslate">
      <span class="pre">
       alignment
      </span>
     </code>
     -  Multiple Tracks in One View​
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#stacked-tracks-in-single-view">
       Stacked tracks in single view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#overlay-tracks-in-single-view">
       Overlay tracks in single view
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#example-gene-annotation">
       Example: Gene Annotation
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#arrangement-arrange-multiple-views">
     <code class="docutils literal notranslate">
      <span class="pre">
       arrangement
      </span>
     </code>
     - Arrange Multiple Views​
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gos-horizontal-gos-vertical">
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.horizontal
        </span>
       </code>
       &amp;
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.vertical
        </span>
       </code>
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#gos-serial-gos-parallel">
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.serial
        </span>
       </code>
       &amp;
       <code class="docutils literal notranslate">
        <span class="pre">
         gos.parallel
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-data">
   Loading data
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-visualization">
     The visualization
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-dataset">
     The dataset
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remote-dataset-via-url">
     Remote dataset (via URL)
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#local-dataset-via-local-filepath">
     Local Dataset (via local filepath)
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#in-memory-via-pd-dataframe">
     In memory (via
     <code class="docutils literal notranslate">
      <span class="pre">
       pd.DataFrame
      </span>
     </code>
     )
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bonus-exercise-local-bigwig-track">
   (BONUS) Exercise: local bigwig track
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="session-2-track-alignment-and-data-loading">
<h1>Session 2: Track Alignment and Data Loading<a class="headerlink" href="#session-2-track-alignment-and-data-loading" title="Permalink to this headline">#</a></h1>
<p>Welcome to the second part of the hands-on tutorial! In this notebook, we:</p>
<ul class="simple">
<li><p>Introduce additional building blocks from the Gosling visualization grammer (track composition primitives)</p></li>
<li><p>Showcase the transparent data-utilties from <strong><code class="docutils literal notranslate"><span class="pre">gos</span></code></strong> to help users visualize their own local (and in-memory) genomics datasets</p></li>
</ul>
<p>To get started, make sure you have <code class="docutils literal notranslate"><span class="pre">gosling</span></code> installed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install gosling<span class="o">[</span>all<span class="o">]==</span><span class="m">0</span>.0.9
<span class="kn">import</span> <span class="nn">gosling</span> <span class="k">as</span> <span class="nn">gos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: gosling[all]==0.0.9 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (0.0.9)
Requirement already satisfied: pandas in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (1.4.3)
Requirement already satisfied: jsonschema&lt;4.0,&gt;=3.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (3.2.0)
Requirement already satisfied: jinja2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (3.1.2)
Requirement already satisfied: starlette in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (0.20.4)
Requirement already satisfied: uvicorn in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (0.18.2)
Requirement already satisfied: portpicker in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (1.5.2)
Requirement already satisfied: gosling-widget in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (0.0.2)
Requirement already satisfied: clodius in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling[all]==0.0.9) (0.19.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: attrs&gt;=17.4.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jsonschema&lt;4.0,&gt;=3.0-&gt;gosling[all]==0.0.9) (21.4.0)
Requirement already satisfied: pyrsistent&gt;=0.14.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jsonschema&lt;4.0,&gt;=3.0-&gt;gosling[all]==0.0.9) (0.18.1)
Requirement already satisfied: six&gt;=1.11.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jsonschema&lt;4.0,&gt;=3.0-&gt;gosling[all]==0.0.9) (1.16.0)
Requirement already satisfied: setuptools in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jsonschema&lt;4.0,&gt;=3.0-&gt;gosling[all]==0.0.9) (58.1.0)
Requirement already satisfied: sortedcontainers in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (2.4.0)
Requirement already satisfied: pydantic in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (1.9.1)
Requirement already satisfied: requests in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (2.28.1)
Requirement already satisfied: cooler&gt;=0.8.5 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (0.8.11)
Requirement already satisfied: Click&gt;=7 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (8.1.3)
Requirement already satisfied: numpy in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (1.23.0)
Requirement already satisfied: negspy in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (0.2.24)
Requirement already satisfied: nose in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (1.3.7)
Requirement already satisfied: pybbi&gt;=0.2.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (0.3.2)
Requirement already satisfied: h5py&gt;=3.0.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (3.7.0)
Requirement already satisfied: pysam in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (0.19.1)
Requirement already satisfied: slugid in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (2.0.0)
Requirement already satisfied: pyfaidx in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (0.7.0)
Requirement already satisfied: dask in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (2022.6.1)
Requirement already satisfied: tqdm in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from clodius-&gt;gosling[all]==0.0.9) (4.64.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pytz&gt;=2020.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from pandas-&gt;gosling[all]==0.0.9) (2022.1)
Requirement already satisfied: python-dateutil&gt;=2.8.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from pandas-&gt;gosling[all]==0.0.9) (2.8.2)
Requirement already satisfied: ipywidgets in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from gosling-widget-&gt;gosling[all]==0.0.9) (7.7.1)
Requirement already satisfied: MarkupSafe&gt;=2.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jinja2-&gt;gosling[all]==0.0.9) (2.1.1)
Requirement already satisfied: psutil in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from portpicker-&gt;gosling[all]==0.0.9) (5.9.1)
Requirement already satisfied: anyio&lt;5,&gt;=3.4.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from starlette-&gt;gosling[all]==0.0.9) (3.6.1)
Requirement already satisfied: h11&gt;=0.8 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from uvicorn-&gt;gosling[all]==0.0.9) (0.13.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: sniffio&gt;=1.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from anyio&lt;5,&gt;=3.4.0-&gt;starlette-&gt;gosling[all]==0.0.9) (1.2.0)
Requirement already satisfied: idna&gt;=2.8 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from anyio&lt;5,&gt;=3.4.0-&gt;starlette-&gt;gosling[all]==0.0.9) (3.3)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pyyaml in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (6.0)
Requirement already satisfied: cytoolz&lt;0.11 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (0.10.1)
Requirement already satisfied: scipy&gt;=0.16 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (1.8.1)
Requirement already satisfied: asciitree in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (0.3.3)
Requirement already satisfied: multiprocess in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (0.70.13)
Requirement already satisfied: pypairix in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (0.3.7)
Requirement already satisfied: simplejson in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (3.17.6)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: toolz&gt;=0.8.2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from dask-&gt;clodius-&gt;gosling[all]==0.0.9) (0.11.2)
Requirement already satisfied: partd&gt;=0.3.10 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from dask-&gt;clodius-&gt;gosling[all]==0.0.9) (1.2.0)
Requirement already satisfied: cloudpickle&gt;=1.1.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from dask-&gt;clodius-&gt;gosling[all]==0.0.9) (2.1.0)
Requirement already satisfied: fsspec&gt;=0.6.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from dask-&gt;clodius-&gt;gosling[all]==0.0.9) (2022.5.0)
Requirement already satisfied: packaging&gt;=20.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from dask-&gt;clodius-&gt;gosling[all]==0.0.9) (21.3)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: ipython-genutils~=0.2.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.2.0)
Requirement already satisfied: jupyterlab-widgets&gt;=1.0.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.1.1)
Requirement already satisfied: widgetsnbextension~=3.6.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (3.6.1)
Requirement already satisfied: ipython&gt;=4.0.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (8.4.0)
Requirement already satisfied: ipykernel&gt;=4.5.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (6.15.0)
Requirement already satisfied: traitlets&gt;=4.3.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (5.3.0)
Requirement already satisfied: typing-extensions&gt;=3.7.4.3 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from pydantic-&gt;clodius-&gt;gosling[all]==0.0.9) (4.3.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: urllib3&lt;1.27,&gt;=1.21.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from requests-&gt;clodius-&gt;gosling[all]==0.0.9) (1.26.9)
Requirement already satisfied: charset-normalizer&lt;3,&gt;=2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from requests-&gt;clodius-&gt;gosling[all]==0.0.9) (2.1.0)
Requirement already satisfied: certifi&gt;=2017.4.17 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from requests-&gt;clodius-&gt;gosling[all]==0.0.9) (2022.6.15)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: tornado&gt;=6.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (6.2)
Requirement already satisfied: matplotlib-inline&gt;=0.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.1.3)
Requirement already satisfied: jupyter-client&gt;=6.1.12 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (7.3.4)
Requirement already satisfied: nest-asyncio in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.5.5)
Requirement already satisfied: debugpy&gt;=1.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.6.0)
Requirement already satisfied: pyzmq&gt;=17 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (23.2.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: backcall in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.2.0)
Requirement already satisfied: stack-data in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.3.0)
Requirement already satisfied: pexpect&gt;4.3 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (4.8.0)
Requirement already satisfied: pygments&gt;=2.4.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (2.12.0)
Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,&lt;3.1.0,&gt;=2.0.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (3.0.30)
Requirement already satisfied: jedi&gt;=0.16 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.18.1)
Requirement already satisfied: decorator in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (5.1.1)
Requirement already satisfied: pickleshare in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.7.5)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pyparsing!=3.0.5,&gt;=2.0.2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from packaging&gt;=20.0-&gt;dask-&gt;clodius-&gt;gosling[all]==0.0.9) (3.0.9)
Requirement already satisfied: locket in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from partd&gt;=0.3.10-&gt;dask-&gt;clodius-&gt;gosling[all]==0.0.9) (1.0.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: notebook&gt;=4.4.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (6.4.12)
Requirement already satisfied: dill&gt;=0.3.5.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from multiprocess-&gt;cooler&gt;=0.8.5-&gt;clodius-&gt;gosling[all]==0.0.9) (0.3.5.1)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: parso&lt;0.9.0,&gt;=0.8.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jedi&gt;=0.16-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.8.3)
Requirement already satisfied: jupyter-core&gt;=4.9.2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jupyter-client&gt;=6.1.12-&gt;ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (4.10.0)
Requirement already satisfied: entrypoints in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from jupyter-client&gt;=6.1.12-&gt;ipykernel&gt;=4.5.1-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.4)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: nbconvert&gt;=5 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (6.5.0)
Requirement already satisfied: Send2Trash&gt;=1.8.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.8.0)
Requirement already satisfied: nbformat in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (5.4.0)
Requirement already satisfied: argon2-cffi in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (21.3.0)
Requirement already satisfied: terminado&gt;=0.8.3 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.15.0)
Requirement already satisfied: prometheus-client in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.14.1)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: ptyprocess&gt;=0.5 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from pexpect&gt;4.3-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.7.0)
Requirement already satisfied: wcwidth in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,&lt;3.1.0,&gt;=2.0.0-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.2.5)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pure-eval in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from stack-data-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.2.2)
Requirement already satisfied: executing in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from stack-data-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.8.3)
Requirement already satisfied: asttokens in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from stack-data-&gt;ipython&gt;=4.0.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (2.0.5)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: nbclient&gt;=0.5.0 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.5.13)
Requirement already satisfied: defusedxml in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.7.1)
Requirement already satisfied: tinycss2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.1.1)
Requirement already satisfied: pandocfilters&gt;=1.4.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.5.0)
Requirement already satisfied: beautifulsoup4 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (4.11.1)
Requirement already satisfied: jupyterlab-pygments in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.2.2)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: bleach in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (5.0.1)
Requirement already satisfied: mistune&lt;2,&gt;=0.8.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.8.4)
Requirement already satisfied: fastjsonschema in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from nbformat-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (2.15.3)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: argon2-cffi-bindings in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from argon2-cffi-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (21.2.0)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: cffi&gt;=1.0.1 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from argon2-cffi-bindings-&gt;argon2-cffi-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (1.15.1)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: soupsieve&gt;1.2 in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from beautifulsoup4-&gt;nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (2.3.2.post1)
Requirement already satisfied: webencodings in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from bleach-&gt;nbconvert&gt;=5-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (0.5.1)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Requirement already satisfied: pycparser in /opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages (from cffi&gt;=1.0.1-&gt;argon2-cffi-bindings-&gt;argon2-cffi-&gt;notebook&gt;=4.4.1-&gt;widgetsnbextension~=3.6.0-&gt;ipywidgets-&gt;gosling-widget-&gt;gosling[all]==0.0.9) (2.21)
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p><strong>Note</strong>: be sure to include the <code class="docutils literal notranslate"><span class="pre">[all]</span></code> for this notebook when installing <code class="docutils literal notranslate"><span class="pre">gos</span></code> from PyPI via <code class="docutils literal notranslate"><span class="pre">pip</span></code>. The special data utilites in <code class="docutils literal notranslate"><span class="pre">gos</span></code> are opt-in by default, and running the install command as shown above ensures that you will have the correct dependencies.</p>
</div></blockquote>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="track-composition">
<h1>Track Composition<a class="headerlink" href="#track-composition" title="Permalink to this headline">#</a></h1>
<p>A <code class="docutils literal notranslate"><span class="pre">gos.Track</span></code> is a unit building block which can be represented as a bar chart, a line chart, ideogram, or other chart types as we’ve seen in the first session. For the concurrent analysis of multiple datasets, multiple <em>Tracks</em> can be grouped into one or more <em>Views</em> and navigated synchronously. Each <code class="docutils literal notranslate"><span class="pre">gos.View</span></code> defines the genomic location for every <code class="docutils literal notranslate"><span class="pre">gos.Track</span></code> it contains, and each <code class="docutils literal notranslate"><span class="pre">gos.Track</span></code> binds and maps data to be visualized.</p>
<p>Changes following View-level properties modifies how <em>Tracks</em> and <em>Views</em> compose together:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">layout</span></code> - Whether to view genomic positions in Cartesian (<code class="docutils literal notranslate"><span class="pre">linear</span></code>) or in polar (<code class="docutils literal notranslate"><span class="pre">circular</span></code>) coordinate systems.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alignment</span></code> - Whether multiple tracks should <code class="docutils literal notranslate"><span class="pre">overlay</span></code> or <code class="docutils literal notranslate"><span class="pre">stack</span></code> within a <em>single</em> View.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">arrangement</span></code> - How to juxtapose <em>multiple</em> Views together (<code class="docutils literal notranslate"><span class="pre">parallel</span></code>, <code class="docutils literal notranslate"><span class="pre">serial</span></code>, <code class="docutils literal notranslate"><span class="pre">vertical</span></code>, <code class="docutils literal notranslate"><span class="pre">horizontal</span></code>).</p></li>
</ul>
<p>This section will walk through each of these properties in more detail with examples.</p>
<section id="layout-linear-and-circular">
<h2><code class="docutils literal notranslate"><span class="pre">layout</span></code> - Linear and Circular<a class="headerlink" href="#layout-linear-and-circular" title="Permalink to this headline">#</a></h2>
<p>The <em>View</em> <code class="docutils literal notranslate"><span class="pre">layout</span></code> property specifies whether genomic coordinates are represented in either a <code class="docutils literal notranslate"><span class="pre">circular</span></code> or <code class="docutils literal notranslate"><span class="pre">linear</span></code> layout.</p>
<p>The following figure displays the upper <em>Track</em> with a <code class="docutils literal notranslate"><span class="pre">linear</span></code> layout and the bottom with a <code class="docutils literal notranslate"><span class="pre">circular</span></code> layout.</p>
<p align="center">
  <img width="600" src="https://github.com/gosling-lang/gosling-website/raw/main/static/img/doc_images/linear_circular.png" />
</p>
<p>Here we create a bar chart visualization displaying a “pseudobulk” excitatory neuron scATAC-seq track from <a class="reference external" href="https://www.nature.com/articles/s41588-020-00721-x">Corces et. al (<em>Nature Genetics, 2020</em>)</a>, and apply a <code class="docutils literal notranslate"><span class="pre">linear</span></code> and <code class="docutils literal notranslate"><span class="pre">circular</span></code> layout.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">bigwig</span><span class="p">(</span>
    <span class="n">url</span><span class="o">=</span><span class="s2">&quot;https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw&quot;</span><span class="p">,</span>
    <span class="n">column</span><span class="o">=</span><span class="s2">&quot;position&quot;</span><span class="p">,</span> 
    <span class="n">value</span><span class="o">=</span><span class="s2">&quot;peak&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">track</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">Track</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span><span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;position:G&quot;</span><span class="p">),</span>
    <span class="n">y</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">Y</span><span class="p">(</span><span class="s2">&quot;peak:Q&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">),</span>
<span class="p">)</span>

<span class="n">track</span><span class="o">.</span><span class="n">view</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-06dbb8a2b79341e4b2ed16405256a293"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-06dbb8a2b79341e4b2ed16405256a293');
    var spec = {"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<section id="linear-view">
<h3>Linear view<a class="headerlink" href="#linear-view" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">track</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">layout</span><span class="o">=</span><span class="s2">&quot;linear&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-f54c7f89cdf64e0bb85a3e5af6e8ed49"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-f54c7f89cdf64e0bb85a3e5af6e8ed49');
    var spec = {"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "layout": "linear"};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="circular-view">
<h3>Circular view<a class="headerlink" href="#circular-view" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">track</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">layout</span><span class="o">=</span><span class="s2">&quot;circular&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-8e7d2671a8d3493cb37daeb5ceb507d5"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-8e7d2671a8d3493cb37daeb5ceb507d5');
    var spec = {"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 200, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "layout": "circular"};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
</section>
<section id="alignment-multiple-tracks-in-one-view">
<h2><code class="docutils literal notranslate"><span class="pre">alignment</span></code> -  Multiple Tracks in One View​<a class="headerlink" href="#alignment-multiple-tracks-in-one-view" title="Permalink to this headline">#</a></h2>
<p>The <em>View</em> <code class="docutils literal notranslate"><span class="pre">alignment</span></code> property allows users to either <code class="docutils literal notranslate"><span class="pre">&quot;overlay&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;stack&quot;</span></code> several tracks.</p>
<p align="center">
  <img width="600" src="https://github.com/gosling-lang/gosling-website/raw/main/static/img/doc_images/alignment.png" />
</p>
<p>When setting <code class="docutils literal notranslate"><span class="pre">alignment</span></code> as <code class="docutils literal notranslate"><span class="pre">&quot;overlay&quot;</span></code>, multiple tracks are layered on top of others. When setting `alignment as “stack”, multiple tracks are vertically concantenated. The default value of alignment is “stack”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">line</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">mark_line</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s2">&quot;#f3b285&quot;</span><span class="p">))</span>

<span class="n">bar</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s2">&quot;#4472c4&quot;</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">bar</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Track({
  color: ColorValue({
    value: &#39;#f3b285&#39;
  }),
  data: {&#39;type&#39;: &#39;bigwig&#39;, &#39;url&#39;: &#39;https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw&#39;, &#39;column&#39;: &#39;position&#39;, &#39;value&#39;: &#39;peak&#39;},
  height: 50,
  mark: &#39;line&#39;,
  width: 800,
  x: X({
    field: &#39;position&#39;,
    type: &#39;genomic&#39;
  }),
  y: Y({
    axis: &#39;right&#39;,
    field: &#39;peak&#39;,
    type: &#39;quantitative&#39;
  })
}) Track({
  color: ColorValue({
    value: &#39;#4472c4&#39;
  }),
  data: {&#39;type&#39;: &#39;bigwig&#39;, &#39;url&#39;: &#39;https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw&#39;, &#39;column&#39;: &#39;position&#39;, &#39;value&#39;: &#39;peak&#39;},
  height: 50,
  mark: &#39;bar&#39;,
  width: 800,
  x: X({
    field: &#39;position&#39;,
    type: &#39;genomic&#39;
  }),
  y: Y({
    axis: &#39;right&#39;,
    field: &#39;peak&#39;,
    type: &#39;quantitative&#39;
  })
})
</pre></div>
</div>
</div>
</div>
<section id="stacked-tracks-in-single-view">
<h3>Stacked tracks in single view<a class="headerlink" href="#stacked-tracks-in-single-view" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">bar</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">spacing</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-d9033f114344477f8408488a9e65484e"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-d9033f114344477f8408488a9e65484e');
    var spec = {"alignment": "stack", "tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "line", "width": 800, "color": {"value": "#f3b285"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "line", "width": 800, "color": {"value": "#f3b285"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "spacing": 1};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span>
    <span class="n">bar</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">),</span>
    <span class="n">line</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">spacing</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;circular&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-d52bd4218a5a478b9dd6ed0af4ddbf22"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-d52bd4218a5a478b9dd6ed0af4ddbf22');
    var spec = {"alignment": "stack", "tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 400, "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "line", "width": 400, "color": {"value": "#f3b285"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "spacing": 1, "layout": "circular"};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="overlay-tracks-in-single-view">
<h3>Overlay tracks in single view<a class="headerlink" href="#overlay-tracks-in-single-view" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">bar</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-86525eb05479485a8f16df6af58d4a58"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-86525eb05479485a8f16df6af58d4a58');
    var spec = {"alignment": "stack", "tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "line", "width": 800, "color": {"value": "#f3b285"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">overlay</span><span class="p">(</span>
    <span class="n">bar</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">),</span>
    <span class="n">line</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">spacing</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;circular&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-1fa7dc59a76e42418a18b32369b131d3"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-1fa7dc59a76e42418a18b32369b131d3');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "mark": "bar", "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "mark": "line", "color": {"value": "#f3b285"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "width": 400, "height": 50, "spacing": 1, "layout": "circular"};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="example-gene-annotation">
<h3>Example: Gene Annotation<a class="headerlink" href="#example-gene-annotation" title="Permalink to this headline">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">overlay</span></code> alignment is an essential feature in Gosling and allows much more complex visual encodings. We leverage this building-block in the code example below to build a gene annotation track, combining several primitive tracks.</p>
<blockquote>
<div><p><strong>Note</strong> its not essential that you understand every detail of the code below. It is provided to demonstrate more complex usage of the <code class="docutils literal notranslate"><span class="pre">gos</span></code> API and can be used for reference later.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gosling</span> <span class="k">as</span> <span class="nn">gos</span>

<span class="c1"># HiGlass-based annotation dataset</span>
<span class="c1"># http://gosling-lang.org/docs/data#beddb-require-higlass-server</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">beddb</span><span class="p">(</span>
    <span class="n">url</span><span class="o">=</span><span class="s2">&quot;https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation&quot;</span><span class="p">,</span>
    <span class="n">genomicFields</span><span class="o">=</span><span class="p">[</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;start&quot;</span><span class="p">},</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;end&quot;</span><span class="p">}</span>
    <span class="p">],</span>
    <span class="n">valueFields</span><span class="o">=</span><span class="p">[</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;nominal&quot;</span><span class="p">},</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;nominal&quot;</span><span class="p">}</span>
    <span class="p">],</span>
    <span class="n">exonIntervalFields</span><span class="o">=</span><span class="p">[</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">12</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;start&quot;</span><span class="p">},</span>
      <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">13</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;end&quot;</span><span class="p">}</span>
    <span class="p">]</span>
<span class="p">)</span>

<span class="c1"># Primitive Tracks</span>

<span class="n">base</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">Track</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">row</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">Row</span><span class="p">(</span><span class="s2">&quot;strand:N&quot;</span><span class="p">,</span> <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]),</span>
    <span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;strand:N&quot;</span><span class="p">,</span> <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">],</span> <span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#7585FF&quot;</span><span class="p">,</span> <span class="s2">&quot;#FF8A85&quot;</span><span class="p">]),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span>
    <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Genes | hg38&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">plusGeneHead</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_triangleRight</span><span class="p">(</span><span class="n">align</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;end:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">size</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">minusGeneHead</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_triangleLeft</span><span class="p">(</span><span class="n">align</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;start:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">size</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">geneLabel</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_text</span><span class="p">(</span><span class="n">dy</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;start:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;end:G&quot;</span><span class="p">,</span>
    <span class="n">text</span><span class="o">=</span><span class="s2">&quot;name:N&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">visibility_lt</span><span class="p">(</span>
    <span class="n">measure</span><span class="o">=</span><span class="s2">&quot;width&quot;</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="s2">&quot;|xe-x|&quot;</span><span class="p">,</span>
    <span class="n">transitionPadding</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">target</span><span class="o">=</span><span class="s2">&quot;mark&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">exon</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_rect</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;start:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;end:G&quot;</span><span class="p">,</span>
    <span class="n">size</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;exon&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">plusGeneRange</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_rule</span><span class="p">(</span><span class="n">linePattern</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;triangleRight&quot;</span><span class="p">,</span> <span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">})</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;start:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;end:G&quot;</span><span class="p">,</span>
    <span class="n">strokeWidth</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">minusGeneRange</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">mark_rule</span><span class="p">(</span><span class="n">linePattern</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;triangleLeft&quot;</span><span class="p">,</span> <span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">})</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;start:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;top&quot;</span><span class="p">),</span>
    <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;end:G&quot;</span><span class="p">,</span>
    <span class="n">strokeWidth</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;type&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
    <span class="n">field</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># Combine tracks with overlay alignment</span>

<span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">overlay</span><span class="p">(</span>
    <span class="n">plusGeneRange</span><span class="p">,</span> <span class="n">minusGeneRange</span><span class="p">,</span> <span class="n">exon</span><span class="p">,</span> <span class="n">plusGeneHead</span><span class="p">,</span> <span class="n">minusGeneHead</span><span class="p">,</span> <span class="n">geneLabel</span>
<span class="p">)</span>

<span class="n">gene_annotation</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span>
    <span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">interval</span><span class="o">=</span><span class="p">[</span><span class="mi">103400000</span><span class="p">,</span> <span class="mi">103700000</span><span class="p">]),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-77a1d265e6624428b562247dc6bc8209"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-77a1d265e6624428b562247dc6bc8209');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleRight", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleLeft", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rect", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["exon"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleRight", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "left"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleLeft", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "right"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "text", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"dy": 15}, "text": {"field": "name", "type": "nominal"}, "title": "Genes | hg38", "visibility": [{"operation": "LT", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10, "target": "mark"}], "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}], "width": 800, "height": 100, "xDomain": {"chromosome": "1", "interval": [103400000, 103700000]}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<p>We can then “stack” our custom <code class="docutils literal notranslate"><span class="pre">gene_annotation</span></code> view with the scATAC-seq track from earlier to provide context.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span>
    <span class="n">gene_annotation</span><span class="p">,</span>
    <span class="n">bar</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Excitatory Neuron ATAC-seq&quot;</span><span class="p">),</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span>
    <span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">interval</span><span class="o">=</span><span class="p">[</span><span class="mi">103400000</span><span class="p">,</span> <span class="mi">103700000</span><span class="p">]),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-665f08a408c6421299fe8803561cfb7a"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-665f08a408c6421299fe8803561cfb7a');
    var spec = {"alignment": "stack", "tracks": [{"alignment": "overlay", "tracks": [{"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleRight", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleLeft", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rect", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["exon"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleRight", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "left"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleLeft", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "right"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "text", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"dy": 15}, "text": {"field": "name", "type": "nominal"}, "title": "Genes | hg38", "visibility": [{"operation": "LT", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10, "target": "mark"}], "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}], "width": 800, "height": 100}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "color": {"value": "#4472c4"}, "title": "Excitatory Neuron ATAC-seq", "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "xDomain": {"chromosome": "1", "interval": [103400000, 103700000]}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
</section>
<section id="arrangement-arrange-multiple-views">
<h2><code class="docutils literal notranslate"><span class="pre">arrangement</span></code> - Arrange Multiple Views​<a class="headerlink" href="#arrangement-arrange-multiple-views" title="Permalink to this headline">#</a></h2>
<p>Gosling supports multi-view visualizations. Users specify the visualizations under the <code class="docutils literal notranslate"><span class="pre">views</span></code> property and modify the arragement of these fields through the <code class="docutils literal notranslate"><span class="pre">arrangement</span></code> property.</p>
<p><code class="docutils literal notranslate"><span class="pre">gos</span></code> provides several top-level utilities to conveniently group multiple views together in each of the arrangement options:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gos.horizontal</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gos.vertical</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gos.serial</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gos.parallel</span></code></p></li>
</ul>
<p align="center">
 <img width="600" src="https://github.com/gosling-lang/gosling-website/raw/main/static/img/doc_images/multi_views.png" />
</p>
<blockquote>
<div><p><strong>Note</strong> the arrangement pairs (<code class="docutils literal notranslate"><span class="pre">gos.serial</span></code>, <code class="docutils literal notranslate"><span class="pre">gos.horizontal</span></code>) and (<code class="docutils literal notranslate"><span class="pre">gos.parallel</span></code>, <code class="docutils literal notranslate"><span class="pre">gos.vertical</span></code>) are <strong>equivalent</strong> when combining <em>Views</em> with <code class="docutils literal notranslate"><span class="pre">linear</span></code> layouts. Behavior <em>only</em> differs when arranging <code class="docutils literal notranslate"><span class="pre">circular</span></code> layout <em>Views</em> (left-most column).</p>
</div></blockquote>
<section id="gos-horizontal-gos-vertical">
<h3><code class="docutils literal notranslate"><span class="pre">gos.horizontal</span></code> &amp; <code class="docutils literal notranslate"><span class="pre">gos.vertical</span></code><a class="headerlink" href="#gos-horizontal-gos-vertical" title="Permalink to this headline">#</a></h3>
<p>We can compose <code class="docutils literal notranslate"><span class="pre">gos.horizontal</span></code> and <code class="docutils literal notranslate"><span class="pre">gos.vertical</span></code> to arrange multiple tracks as into separate views. Notice how interactions are no longer synchronized since each <em>Track</em> is within a unique <em>View</em>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">horizontal</span><span class="p">(</span>
    <span class="c1"># left</span>
    <span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span>
        <span class="n">track</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
        <span class="n">track</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
    <span class="p">),</span>
    <span class="c1"># right</span>
    <span class="n">track</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s2">&quot;#4472c4&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">240</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-e893c24de3df45b780b3e08ed57e23a7"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-e893c24de3df45b780b3e08ed57e23a7');
    var spec = {"arrangement": "horizontal", "views": [{"alignment": "stack", "tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 100, "mark": "bar", "width": 300, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 100, "mark": "bar", "width": 300, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]}, {"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 240, "mark": "bar", "width": 600, "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]}]};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<p>We can selectively “group” tracks within the <strong>same</strong> view using the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> features in the previous section. We replace <code class="docutils literal notranslate"><span class="pre">gos.vertical</span></code> with <code class="docutils literal notranslate"><span class="pre">gos.stack</span></code>, which combines the tracks on the left into the same <em>View</em>.”</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">horizontal</span><span class="p">(</span>
    <span class="c1"># left</span>
    <span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span>
        <span class="n">gene_annotation</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
        <span class="n">track</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
    <span class="p">),</span>
    <span class="c1"># right</span>
    <span class="n">track</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s2">&quot;#4472c4&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">240</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-fea530637caf4957842c046d595f3537"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-fea530637caf4957842c046d595f3537');
    var spec = {"arrangement": "horizontal", "views": [{"alignment": "stack", "tracks": [{"alignment": "overlay", "tracks": [{"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleRight", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleLeft", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rect", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["exon"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleRight", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "left"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleLeft", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "right"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "text", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"dy": 15}, "text": {"field": "name", "type": "nominal"}, "title": "Genes | hg38", "visibility": [{"operation": "LT", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10, "target": "mark"}], "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}], "width": 300, "height": 100}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 100, "mark": "bar", "width": 300, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]}, {"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 240, "mark": "bar", "width": 600, "color": {"value": "#4472c4"}, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}]}]};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="gos-serial-gos-parallel">
<h3><code class="docutils literal notranslate"><span class="pre">gos.serial</span></code> &amp; <code class="docutils literal notranslate"><span class="pre">gos.parallel</span></code><a class="headerlink" href="#gos-serial-gos-parallel" title="Permalink to this headline">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">serial</span></code> and <code class="docutils literal notranslate"><span class="pre">parallel</span></code> alignments differ in behavior from <code class="docutils literal notranslate"><span class="pre">horizontal</span></code> and <code class="docutils literal notranslate"><span class="pre">veritcal</span></code> when combining <em>Views</em> with <code class="docutils literal notranslate"><span class="pre">circular</span></code> alignments.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gos</span><span class="o">.</span><span class="n">serial</span><span class="p">(</span>
    <span class="n">track</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">layout</span><span class="o">=</span><span class="s2">&quot;circular&quot;</span><span class="p">),</span> 
    <span class="n">gene_annotation</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;circular&quot;</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-90af46ab218343379b45f2f0606cda7b"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-90af46ab218343379b45f2f0606cda7b');
    var spec = {"arrangement": "serial", "views": [{"tracks": [{"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 300, "x": {"field": "position", "type": "genomic"}, "y": {"axis": "right", "field": "peak", "type": "quantitative"}}], "layout": "circular"}, {"alignment": "overlay", "tracks": [{"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleRight", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleLeft", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rect", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["exon"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleRight", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "left"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleLeft", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "right"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "text", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"dy": 15}, "text": {"field": "name", "type": "nominal"}, "title": "Genes | hg38", "visibility": [{"operation": "LT", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10, "target": "mark"}], "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}], "width": 300, "height": 100, "layout": "circular"}]};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="loading-data">
<h1>Loading data<a class="headerlink" href="#loading-data" title="Permalink to this headline">#</a></h1>
<p>This section illustrates a key (optional) feature of <code class="docutils literal notranslate"><span class="pre">gos</span></code> which makes hosting data for your Gosling visualizations a breeze.</p>
<p>Normally a Gosling visualization requires the administration of a web-server to host both the client and genomics data sets for the visualization. In <code class="docutils literal notranslate"><span class="pre">gos</span></code>, we provide further integration with Python to hide this complexity and allow remote, local, and in-memory data to be visualized seamlessly through an idential API.</p>
<p>In this notebook, we will visualize the same <a class="reference external" href="https://samtools.github.io/hts-specs/BEDv1.pdf">BED file</a> containing h38 cytoband information as a:</p>
<ul class="simple">
<li><p>remote dataset (via URL)</p></li>
<li><p>local dataset (via local path)</p></li>
<li><p>in memory (from a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>)</p></li>
</ul>
<section id="the-visualization">
<h2>The visualization<a class="headerlink" href="#the-visualization" title="Permalink to this headline">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">ideogram</span></code> function generates an ideogram visualization for a given Gosling data source. It is not important that you understand the details of this block to follow along in this notebook. Moreover, the important bit is to understand that <code class="docutils literal notranslate"><span class="pre">ideogram</span></code> takes <code class="docutils literal notranslate"><span class="pre">data</span></code> as input and returns a Gosling visualization created with the <code class="docutils literal notranslate"><span class="pre">gos</span></code> API.</p>
<p>We will show how this function can be <em>reused</em> for various <code class="docutils literal notranslate"><span class="pre">data</span></code> defintions (genomic data sources).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">gos</span><span class="o">.</span><span class="n">View</span><span class="p">:</span>
    <span class="n">track</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">Track</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="c1"># bind data to track</span>
    
    <span class="n">arms</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">mark_rect</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;stain:N&quot;</span><span class="p">,</span>
            <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gneg&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos25&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos50&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos75&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos100&quot;</span><span class="p">,</span> <span class="s2">&quot;gvar&quot;</span><span class="p">],</span>
            <span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="s2">&quot;#D9D9D9&quot;</span><span class="p">,</span> <span class="s2">&quot;#979797&quot;</span><span class="p">,</span> <span class="s2">&quot;#636363&quot;</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;#A0A0F2&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;chromStart:G&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">),</span>
        <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;chromEnd:G&quot;</span><span class="p">,</span>
        <span class="n">stroke</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s2">&quot;black&quot;</span><span class="p">),</span>
        <span class="n">strokeWidth</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mf">0.5</span><span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">transform_filter_not</span><span class="p">(</span>
        <span class="n">field</span><span class="o">=</span><span class="s2">&quot;stain&quot;</span><span class="p">,</span>
        <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;acen&quot;</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="n">labels</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">mark_text</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">text</span><span class="o">=</span><span class="s2">&quot;name:N&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;stain:N&quot;</span><span class="p">,</span>
            <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gneg&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos25&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos50&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos75&quot;</span><span class="p">,</span> <span class="s2">&quot;gpos100&quot;</span><span class="p">,</span> <span class="s2">&quot;gvar&quot;</span><span class="p">],</span>
            <span class="nb">range</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;#636363&quot;</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;#D9D9D9&quot;</span><span class="p">,</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="n">strokeWidth</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="p">)</span><span class="o">.</span><span class="n">visibility_lt</span><span class="p">(</span>
        <span class="n">target</span><span class="o">=</span><span class="s1">&#39;mark&#39;</span><span class="p">,</span>
        <span class="n">measure</span><span class="o">=</span><span class="s1">&#39;width&#39;</span><span class="p">,</span>
        <span class="n">threshold</span><span class="o">=</span><span class="s1">&#39;|xe-x|&#39;</span><span class="p">,</span>
        <span class="n">transitionPadding</span><span class="o">=</span><span class="mi">10</span>
    <span class="p">)</span>

    <span class="n">centromere</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;chromStart:G&quot;</span><span class="p">),</span>
        <span class="n">xe</span><span class="o">=</span><span class="s2">&quot;chromEnd:G&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="s1">&#39;red&#39;</span><span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
        <span class="s2">&quot;stain&quot;</span><span class="p">,</span> <span class="n">oneOf</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;acen&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="n">centromere_left</span> <span class="o">=</span> <span class="n">centromere</span><span class="o">.</span><span class="n">mark_triangleLeft</span><span class="p">()</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
        <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="s2">&quot;p&quot;</span>
    <span class="p">)</span>

    <span class="n">centromere_right</span> <span class="o">=</span> <span class="n">centromere</span><span class="o">.</span><span class="n">mark_triangleRight</span><span class="p">()</span><span class="o">.</span><span class="n">transform_filter</span><span class="p">(</span>
        <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="s2">&quot;q&quot;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">gos</span><span class="o">.</span><span class="n">overlay</span><span class="p">(</span><span class="n">arms</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">centromere_left</span><span class="p">,</span> <span class="n">centromere_right</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">height</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="the-dataset">
<h2>The dataset<a class="headerlink" href="#the-dataset" title="Permalink to this headline">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">url</span></code> below links to a <a class="reference external" href="https://samtools.github.io/hts-specs/BEDv1.pdf">BED4+1</a> file containing UCSC hg38 cytoband information. This dataset is hosted on GitHub and is avaiable via URL.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>curl -s <span class="o">{</span>url<span class="o">}</span>  <span class="p">|</span> head <span class="p">|</span> column -t
<span class="c1"># chrom  chromStart  chromEnd  name  stain</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr1  0         2300000   p36.33  gneg
chr1  2300000   5300000   p36.32  gpos25
chr1  5300000   7100000   p36.31  gneg
chr1  7100000   9100000   p36.23  gpos25
chr1  9100000   12500000  p36.22  gneg
chr1  12500000  15900000  p36.21  gpos50
chr1  15900000  20100000  p36.13  gneg
chr1  20100000  23600000  p36.12  gpos25
chr1  23600000  27600000  p36.11  gneg
chr1  27600000  29900000  p35.3   gpos25
</pre></div>
</div>
</div>
</div>
</section>
<section id="remote-dataset-via-url">
<h2>Remote dataset (via URL)<a class="headerlink" href="#remote-dataset-via-url" title="Permalink to this headline">#</a></h2>
<p>We can reference this URL directly in Gos by creating a CSV data source via <code class="docutils literal notranslate"><span class="pre">gos.csv(...)</span></code>. This function returns a Python dictionary that describes our dataset to Gosling. We use the <code class="docutils literal notranslate"><span class="pre">gos.csv</span></code> utility since the resource is a columnar text dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># specify BED4+1 format</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
    <span class="n">url</span><span class="o">=</span><span class="n">url</span><span class="p">,</span>
    <span class="n">headerNames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;stain&#39;</span><span class="p">],</span> <span class="c1"># the +1 field is stain</span>
    <span class="n">chromosomeField</span><span class="o">=</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="c1"># the column containing chrom names</span>
    <span class="n">genomicFields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">],</span> <span class="c1"># fields with genomic coordinates</span>
    <span class="n">separator</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;type&#39;: &#39;csv&#39;,
 &#39;url&#39;: &#39;https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed&#39;,
 &#39;headerNames&#39;: [&#39;chrom&#39;, &#39;chromStart&#39;, &#39;chromEnd&#39;, &#39;name&#39;, &#39;stain&#39;],
 &#39;chromosomeField&#39;: &#39;chrom&#39;,
 &#39;genomicFields&#39;: [&#39;chromStart&#39;, &#39;chromEnd&#39;],
 &#39;separator&#39;: &#39;\t&#39;}
</pre></div>
</div>
</div>
</div>
<p>We can now pass this dataset directly to the <code class="docutils literal notranslate"><span class="pre">ideogram</span></code> function which binds <code class="docutils literal notranslate"><span class="pre">data</span></code> to <code class="docutils literal notranslate"><span class="pre">gos.Track</span></code> and creates our custom visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-26f32e0b49e54e408e766f4a347c7b4f"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-26f32e0b49e54e408e766f4a347c7b4f');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "rect", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["white", "#D9D9D9", "#979797", "#636363", "black", "#A0A0F2"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"], "not": true}], "stroke": {"value": "black"}, "strokeWidth": {"value": 0.5}, "x": {"axis": "none", "field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "text", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["black", "#636363", "black", "#D9D9D9", "white", "black"], "type": "nominal"}, "strokeWidth": {"value": 0}, "text": {"field": "name", "type": "nominal"}, "visibility": [{"operation": "LT", "target": "mark", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10}]}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleLeft", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "p"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleRight", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "q"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}], "width": 800, "height": 20};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<p>This visualization is a bit crowded since we are viewing the data genome-wide. We can set the initial genomic domain for the visualization to Chromosome 2 by specifying <code class="docutils literal notranslate"><span class="pre">xDomain</span></code> as a property.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;chr2&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-4b2e3c64f09044d69cda455c2098d81d"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-4b2e3c64f09044d69cda455c2098d81d');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "rect", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["white", "#D9D9D9", "#979797", "#636363", "black", "#A0A0F2"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"], "not": true}], "stroke": {"value": "black"}, "strokeWidth": {"value": 0.5}, "x": {"axis": "none", "field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "text", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["black", "#636363", "black", "#D9D9D9", "white", "black"], "type": "nominal"}, "strokeWidth": {"value": 0}, "text": {"field": "name", "type": "nominal"}, "visibility": [{"operation": "LT", "target": "mark", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10}]}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleLeft", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "p"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleRight", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "q"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}], "width": 800, "height": 20, "xDomain": {"chromosome": "chr2"}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="local-dataset-via-local-filepath">
<h2>Local Dataset (via local filepath)<a class="headerlink" href="#local-dataset-via-local-filepath" title="Permalink to this headline">#</a></h2>
<p>Data are not always publically available via URL like above, and often we’d like to visualize local data files. To visualize local data, <strong>simply change the URL to a local file path</strong>.</p>
<div class="highlight-diff notranslate"><div class="highlight"><pre><span></span><span class="w">data = gos.csv(</span>
<span class="gd">-  url=url,</span><span class="w"></span>
<span class="gi">+  url=&quot;./UCSC.HG38.Human.CytoBandIdeogram.bed&quot;,</span><span class="w"></span>
<span class="w"> </span>  ... <span class="w"></span>
<span class="w">)</span>
</pre></div>
</div>
<p>Below we download the file from GitHub and load the visualization from our local filesytem.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget <span class="o">{</span>url<span class="o">}</span> # download file
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2022-07-06 01:53:18--  https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 30748 (30K) [text/plain]
Saving to: ‘UCSC.HG38.Human.CytoBandIdeogram.bed’


          UCSC.HG38   0%[                    ]       0  --.-KB/s               
UCSC.HG38.Human.Cyt 100%[===================&gt;]  30.03K  --.-KB/s    in 0s      

2022-07-06 01:53:18 (87.7 MB/s) - ‘UCSC.HG38.Human.CytoBandIdeogram.bed’ saved [30748/30748]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>cat UCSC.HG38.Human.CytoBandIdeogram.bed <span class="p">|</span> head <span class="p">|</span> column -t # print <span class="nb">local</span> file contents
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr1  0         2300000   p36.33  gneg
chr1  2300000   5300000   p36.32  gpos25
chr1  5300000   7100000   p36.31  gneg
chr1  7100000   9100000   p36.23  gpos25
chr1  9100000   12500000  p36.22  gneg
chr1  12500000  15900000  p36.21  gpos50
chr1  15900000  20100000  p36.13  gneg
chr1  20100000  23600000  p36.12  gpos25
chr1  23600000  27600000  p36.11  gneg
chr1  27600000  29900000  p35.3   gpos25
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>ls
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>01-Single-Track.ipynb			   04-Coordinated-Multiple-Views.ipynb
02-Track-Alignment-and-Data-Loading.ipynb  Index.ipynb
03-Semantic-Zooming.ipynb		   UCSC.HG38.Human.CytoBandIdeogram.bed
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
    <span class="n">url</span><span class="o">=</span><span class="s2">&quot;./UCSC.HG38.Human.CytoBandIdeogram.bed&quot;</span><span class="p">,</span>
    <span class="c1"># url=url</span>
    <span class="n">headerNames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;stain&#39;</span><span class="p">],</span>
    <span class="n">chromosomeField</span><span class="o">=</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
    <span class="n">genomicFields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">],</span>
    <span class="n">separator</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># reuse the same visualization</span>
<span class="n">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;chr2&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-0fba642fd74a434a888feefabd5a732c"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-0fba642fd74a434a888feefabd5a732c');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "csv", "url": "http://localhost:45621/3148a7d4-45b3-4ef1-a6bd-770d7fee4e9b.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "rect", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["white", "#D9D9D9", "#979797", "#636363", "black", "#A0A0F2"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"], "not": true}], "stroke": {"value": "black"}, "strokeWidth": {"value": 0.5}, "x": {"axis": "none", "field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/3148a7d4-45b3-4ef1-a6bd-770d7fee4e9b.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "text", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["black", "#636363", "black", "#D9D9D9", "white", "black"], "type": "nominal"}, "strokeWidth": {"value": 0}, "text": {"field": "name", "type": "nominal"}, "visibility": [{"operation": "LT", "target": "mark", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10}]}, {"data": {"type": "csv", "url": "http://localhost:45621/3148a7d4-45b3-4ef1-a6bd-770d7fee4e9b.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleLeft", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "p"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/3148a7d4-45b3-4ef1-a6bd-770d7fee4e9b.bed", "headerNames": ["chrom", "chromStart", "chromEnd", "name", "stain"], "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"], "separator": "\t"}, "mark": "triangleRight", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "q"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}], "width": 800, "height": 20, "xDomain": {"chromosome": "chr2"}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
<section id="in-memory-via-pd-dataframe">
<h2>In memory (via <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>)<a class="headerlink" href="#in-memory-via-pd-dataframe" title="Permalink to this headline">#</a></h2>
<p>While loading remote and local genomics data files is useful, often we want to visualize intermediate or derived information during analysis. Rather than writing these results to disk, <code class="docutils literal notranslate"><span class="pre">gos</span></code> supports visualizing in-memory data directly from Pandas dataframes <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>.</p>
<p>In order to use this feature, we first load our dataset as a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s1">&#39;./UCSC.HG38.Human.CytoBandIdeogram.bed&#39;</span><span class="p">,</span> 
    <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;stain&#39;</span><span class="p">],</span>
    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
<span class="p">)</span>

<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>chromStart</th>
      <th>chromEnd</th>
      <th>name</th>
      <th>stain</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>0</td>
      <td>2300000</td>
      <td>p36.33</td>
      <td>gneg</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>2300000</td>
      <td>5300000</td>
      <td>p36.32</td>
      <td>gpos25</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>5300000</td>
      <td>7100000</td>
      <td>p36.31</td>
      <td>gneg</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>7100000</td>
      <td>9100000</td>
      <td>p36.23</td>
      <td>gpos25</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>9100000</td>
      <td>12500000</td>
      <td>p36.22</td>
      <td>gneg</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Lets filter <code class="docutils literal notranslate"><span class="pre">df</span></code> in Python for our dataset only contains entries for Chromosome 2.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr2&quot;</span><span class="p">]</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>chromStart</th>
      <th>chromEnd</th>
      <th>name</th>
      <th>stain</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>63</th>
      <td>chr2</td>
      <td>0</td>
      <td>4400000</td>
      <td>p25.3</td>
      <td>gneg</td>
    </tr>
    <tr>
      <th>64</th>
      <td>chr2</td>
      <td>4400000</td>
      <td>6900000</td>
      <td>p25.2</td>
      <td>gpos50</td>
    </tr>
    <tr>
      <th>65</th>
      <td>chr2</td>
      <td>6900000</td>
      <td>12000000</td>
      <td>p25.1</td>
      <td>gneg</td>
    </tr>
    <tr>
      <th>66</th>
      <td>chr2</td>
      <td>12000000</td>
      <td>16500000</td>
      <td>p24.3</td>
      <td>gpos75</td>
    </tr>
    <tr>
      <th>67</th>
      <td>chr2</td>
      <td>16500000</td>
      <td>19000000</td>
      <td>p24.2</td>
      <td>gneg</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can now create a <code class="docutils literal notranslate"><span class="pre">data</span></code> source for our visualization using the <code class="docutils literal notranslate"><span class="pre">df.gos.csv(...)</span></code> method, and visualize directly! Note how the resulting visualization only renders for chromosome 2.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">gos</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>    <span class="c1"># we only need to specify these fields since the rest are inferred from dataframe</span>
    <span class="n">chromosomeField</span><span class="o">=</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
    <span class="n">genomicFields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">],</span> 
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;type&#39;: &#39;csv&#39;,
 &#39;url&#39;: &#39;http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv&#39;,
 &#39;chromosomeField&#39;: &#39;chrom&#39;,
 &#39;genomicFields&#39;: [&#39;chromStart&#39;, &#39;chromEnd&#39;]}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">gos</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
    <span class="c1"># we only need to specify these fields since the rest are inferred from dataframe</span>
    <span class="n">chromosomeField</span><span class="o">=</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
    <span class="n">genomicFields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromStart&quot;</span><span class="p">,</span> <span class="s2">&quot;chromEnd&quot;</span><span class="p">],</span> 
<span class="p">)</span>

<span class="n">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="c1"># view in context of full assembly</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-e58bd893de9b427897ff0efb75459926"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-e58bd893de9b427897ff0efb75459926');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "rect", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["white", "#D9D9D9", "#979797", "#636363", "black", "#A0A0F2"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"], "not": true}], "stroke": {"value": "black"}, "strokeWidth": {"value": 0.5}, "x": {"axis": "none", "field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "text", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["black", "#636363", "black", "#D9D9D9", "white", "black"], "type": "nominal"}, "strokeWidth": {"value": 0}, "text": {"field": "name", "type": "nominal"}, "visibility": [{"operation": "LT", "target": "mark", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10}]}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "triangleLeft", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "p"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "triangleRight", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "q"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}], "width": 800, "height": 20};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ideogram</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;chr2&quot;</span><span class="p">))</span> <span class="c1"># view just chrom 2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-fbacf52acab54329aab9c2278d9a9102"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-fbacf52acab54329aab9c2278d9a9102');
    var spec = {"alignment": "overlay", "tracks": [{"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "rect", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["white", "#D9D9D9", "#979797", "#636363", "black", "#A0A0F2"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"], "not": true}], "stroke": {"value": "black"}, "strokeWidth": {"value": 0.5}, "x": {"axis": "none", "field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "text", "color": {"domain": ["gneg", "gpos25", "gpos50", "gpos75", "gpos100", "gvar"], "field": "stain", "range": ["black", "#636363", "black", "#D9D9D9", "white", "black"], "type": "nominal"}, "strokeWidth": {"value": 0}, "text": {"field": "name", "type": "nominal"}, "visibility": [{"operation": "LT", "target": "mark", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10}]}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "triangleLeft", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "p"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}, {"data": {"type": "csv", "url": "http://localhost:45621/582136adca397a8e7c5e78380456fc5a.csv", "chromosomeField": "chrom", "genomicFields": ["chromStart", "chromEnd"]}, "mark": "triangleRight", "color": {"value": "red"}, "dataTransform": [{"type": "filter", "field": "stain", "oneOf": ["acen"]}, {"type": "filter", "field": "name", "include": "q"}], "x": {"field": "chromStart", "type": "genomic"}, "xe": {"field": "chromEnd", "type": "genomic"}}], "width": 800, "height": 20, "xDomain": {"chromosome": "chr2"}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="bonus-exercise-local-bigwig-track">
<h1>(BONUS) Exercise: local bigwig track<a class="headerlink" href="#bonus-exercise-local-bigwig-track" title="Permalink to this headline">#</a></h1>
<p>In the previous notebook we visualized several scATAC-seq “pseudobulk” tracks from human brain cells using a simple barplot.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">bigwig_url</span></code> below points to a <a class="reference external" href="https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw">BigWig file (161Mb)</a> on a <strong>remote</strong> server containing the normliazed aggregate signal from excitatory neurons cells in the human brain.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bigwig_url</span> <span class="o">=</span> <span class="s2">&quot;https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>We can combine the <code class="docutils literal notranslate"><span class="pre">gene_annotation</span></code> <em>View</em> from earlier with a track for our scATAC-seq dataset to provide context when visualizing the peaks.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_bw</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">bigwig</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="n">bigwig_url</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="s2">&quot;peak&quot;</span><span class="p">)</span>

<span class="n">excitatory_neurons</span> <span class="o">=</span> <span class="n">gos</span><span class="o">.</span><span class="n">Track</span><span class="p">(</span><span class="n">data_bw</span><span class="p">)</span><span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;position:G&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;peak:Q&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span><span class="n">height</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>

<span class="n">gos</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">,</span> <span class="n">excitatory_neurons</span><span class="p">)</span><span class="o">.</span><span class="n">properties</span><span class="p">(</span>
    <span class="n">xDomain</span><span class="o">=</span><span class="n">gos</span><span class="o">.</span><span class="n">GenomicDomain</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">interval</span><span class="o">=</span><span class="p">[</span><span class="mi">103400000</span><span class="p">,</span> <span class="mi">103700000</span><span class="p">]),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<!DOCTYPE html>
<html>
<head>
  <style>.error { color: red; }</style>
  <link rel="stylesheet" href="https://unpkg.com/higlass@1.11/dist/hglib.css">
</head>
<body>
  <div id="jupyter-gosling-8a30704c0a6d4b9bbcdcc9d39b032de5"></div>
  <script type="module">

    async function loadScript(src) {
        return new Promise(resolve => {
            const script = document.createElement('script');
            script.onload = resolve;
            script.src = src;
            script.async = false;
            document.head.appendChild(script);
        });
    }

    async function loadGosling() {
        // Manually load scripts from window namespace since requirejs might not be
        // available in all browser environments.

        // https://github.com/DanielHreben/requirejs-toggle
        if (!window.gosling) {

            // https://github.com/DanielHreben/requirejs-toggle
            window.__requirejsToggleBackup = {
                define: window.define,
                require: window.require,
                requirejs: window.requirejs,
            };

            for (const field of Object.keys(window.__requirejsToggleBackup)) {
                window[field] = undefined;
            }

            // load dependencies sequentially
            const sources = [
                "https://unpkg.com/react@17/umd/react.production.min.js",
                "https://unpkg.com/react-dom@17/umd/react-dom.production.min.js",
                "https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js",
                "https://unpkg.com/higlass@1.11/dist/hglib.js",
                "https://unpkg.com/gosling.js@0.9.20/dist/gosling.js",
            ];

            for (const src of sources) await loadScript(src);

            // restore requirejs after scripts have loaded
            Object.assign(window, window.__requirejsToggleBackup);
            delete window.__requirejsToggleBackup;

        }

        return window.gosling;
    };

    var el = document.getElementById('jupyter-gosling-8a30704c0a6d4b9bbcdcc9d39b032de5');
    var spec = {"alignment": "stack", "tracks": [{"alignment": "overlay", "tracks": [{"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleRight", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rule", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "strokeWidth": {"value": 3}, "style": {"linePattern": {"type": "triangleLeft", "size": 5}}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "rect", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["exon"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleRight", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["+"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "left"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "end", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "triangleLeft", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}, {"type": "filter", "field": "strand", "oneOf": ["-"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"align": "right"}, "title": "Genes | hg38", "x": {"axis": "top", "field": "start", "type": "genomic"}}, {"data": {"type": "beddb", "url": "https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation", "genomicFields": [{"index": 1, "name": "start"}, {"index": 2, "name": "end"}], "valueFields": [{"index": 5, "name": "strand", "type": "nominal"}, {"index": 3, "name": "name", "type": "nominal"}], "exonIntervalFields": [{"index": 12, "name": "start"}, {"index": 13, "name": "end"}]}, "mark": "text", "color": {"domain": ["+", "-"], "field": "strand", "range": ["#7585FF", "#FF8A85"], "type": "nominal"}, "dataTransform": [{"type": "filter", "field": "type", "oneOf": ["gene"]}], "row": {"domain": ["+", "-"], "field": "strand", "type": "nominal"}, "size": {"value": 15}, "style": {"dy": 15}, "text": {"field": "name", "type": "nominal"}, "title": "Genes | hg38", "visibility": [{"operation": "LT", "measure": "width", "threshold": "|xe-x|", "transitionPadding": 10, "target": "mark"}], "x": {"axis": "top", "field": "start", "type": "genomic"}, "xe": {"field": "end", "type": "genomic"}}], "width": 800, "height": 100}, {"data": {"type": "bigwig", "url": "https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw", "column": "position", "value": "peak"}, "height": 50, "mark": "bar", "width": 800, "x": {"field": "position", "type": "genomic"}, "y": {"field": "peak", "type": "quantitative"}}], "xDomain": {"chromosome": "1", "interval": [103400000, 103700000]}};
    var opt = {"padding": 0, "theme": null};

    loadGosling()
        .then(gosling => gosling.embed(el, spec, opt))
        .catch(err => {
            el.innerHTML = `<div class="error">
    <p>JavaScript Error: ${error.message}</p>
    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>
</div>`;
            throw error;
        });
  </script>
</body>
</html></div></div>
</div>
<p>Data is not always available via public URL, and often will be stored locally during an analysis.</p>
<p><strong>Task</strong>:</p>
<ol class="simple">
<li><p><em>download</em> the scATAC-seq dataset locally (<strong>hint</strong>: use <code class="docutils literal notranslate"><span class="pre">!wget</span></code> in the cell below)</p></li>
<li><p><em>change</em> the code snippet above to load the <em>local</em> <code class="docutils literal notranslate"><span class="pre">ExcitatoryNeurons-insertions_bin100_RIPnorm.bw</span></code></p></li>
</ol>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="01-Single-Track.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Session 1: Single Track</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="03-Semantic-Zooming.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Session 3: Semantic Zooming</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Sehi L'Yi, Trevor Manz, Qianwen Wang, Nils Gehlenborg<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>